modules/rostests/kmtests/kmtest/service.c

```diff
@@ -101,6 +102,102 @@ KmtCreateService(
                              ServiceHandle);
 }
 
+/**
+ * @name KmtGetServiceStateAsString
+ *
+ * @param ServiceState
+ *        Service state as a number
+ *
+ * @return Service state as a string
+ */
+static
+PCSTR
+KmtGetServiceStateAsString(
+    IN DWORD ServiceState)
+{
+    switch(ServiceState)
+    {
+        case SERVICE_STOPPED:
+            return "STOPPED";
+        case SERVICE_START_PENDING:
+            return "START_PENDING";
+        case SERVICE_STOP_PENDING:
+            return "STOP_PENDING";
+        case SERVICE_RUNNING:
+            return "RUNNING";
+        case SERVICE_CONTINUE_PENDING:
+            return "CONTINUE_PENDING";
+        case SERVICE_PAUSE_PENDING:
+            return "PAUSE_PENDING";
+        case SERVICE_PAUSED:
+            return "PAUSED";
+        default:
+            return "(Unknown)";
+    }
+}
+
+/**
+ * @name KmtCheckServiceState
+ *
+ * @param ServiceName
+ *        Name of the service to check,
+ *        or NULL
+ * @param ServiceHandle
+ *        Handle to the service
+ * @param ExpectedServiceState
+ *        State which the service should be in
+ *
+ * @return Win32 error code
+ */
+static
+DWORD
+KmtCheckServiceState(
+    IN PCWSTR ServiceName OPTIONAL,
+    IN SC_HANDLE ServiceHandle,
+    IN DWORD ExpectedServiceState)
+{
+    DWORD Error = ERROR_SUCCESS;
+    SERVICE_STATUS ServiceStatus;
+    DWORD StartTime = GetTickCount();
+    DWORD Timeout = 10 * 1000;
+    PCWSTR ServiceNameOut = ServiceName ? ServiceName : L"(handle only, no name)";
+
+    assert(ServiceHandle);
+    assert(ExpectedServiceState);
+
+    if (!QueryServiceStatus(ServiceHandle, &ServiceStatus))
+        error_goto(Error, cleanup);
+
+    while (ServiceStatus.dwCurrentState != ExpectedServiceState)
+    {
+        // NB: Not bothering with ServiceStatus.dwWaitHint nor ServiceStatus.dwCheckPoint.
```

Comment:
Not sure I understand this comment. "bothering" to do it is less code than the length of this comment.